### PR TITLE
Add module: nf-core hackathon #207 : new module: msisensor/msi

### LINKED ISSUE #207: new module: msisensor/msi
I think it would be good to have module for [msisensor](https://bioconda.github.io/recipes/msisensor/README.html)

### PR DESCRIPTION
**Pull request by kevbrick**
NOTES:
Addition of msisensor/msi module necessitated the addition of a second module (msisensor/scan) - also included in this pull request. Discussed on nf-core slack and decided that two modules is better. msisensor/msi requires the output file from msisensor/scan. 

TESTS: 
The tests for this module are not ideal - they test that the tool will run but DO NOT test that the tool returns sensible values. This is because the tool is designed to work with a tumor/normal pair of BAMs, but SARS/Cov2 BAMs are used for testing. Room for improvement once more relevant nf-core test data come available.  

## PR checklist

Closes #207 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: software/msisensor/msi/main.nf
Comment (grst) [inline software/msisensor/msi/main.nf:25]: this generates a tuple with `[val, path, path, path, path, path]`. 
If you want to only emit a single `report` rather than each output type in a separate channel, 
I'd suggest to use a wildcard to just generate a `[val, path]` tuple:

```
path("*.{output,output_dis,output_germline,output_somatic}")
```

File: software/msisensor/msi/main.nf
Comment (grst) [inline software/msisensor/msi/main.nf:32]: I don't think we need that anymore?

File: software/msisensor/msi/main.nf
Comment (grst) [inline software/msisensor/msi/main.nf:40]: would be great if you could get this to work! If it is absolutely impossible to retrieve the version number from the program itself, we usually use the pattern to `def VERSION=xxx` at the top of the file and then write that constant into the `txt` file.

File: tests/software/msisensor/msi/main.nf
Comment (grst) [inline tests/software/msisensor/msi/main.nf:26]: Can you please update this to use the new params config file?

**Review by grst — CHANGES_REQUESTED** [review]
Thanks for adding this! 

Apart from the comments, can you please add an entry to the `pytest_software.yml` file - it is required to trigger the tests on the CI.

File: software/msisensor/msi/main.nf
Comment (kevbrick) [reply software/msisensor/msi/main.nf:25]: :+1:

File: software/msisensor/msi/main.nf
Comment (kevbrick) [reply software/msisensor/msi/main.nf:32]: :smile: oops ... fixed in update

File: software/msisensor/msi/main.nf
Comment (kevbrick) [reply software/msisensor/msi/main.nf:40]: yep. Fixed

File: tests/software/msisensor/msi/main.nf
Comment (kevbrick) [reply tests/software/msisensor/msi/main.nf:26]: fixed in new commit

**Review by kevbrick — COMMENTED** [review]
Fixed issues in code review and modified to use new test data config.

File: software/msisensor/msi/main.nf
Comment (grst) [inline software/msisensor/msi/main.nf:22]: maybe this should be split up into separate channels? I'd love to hear a second opinion on that

File: software/msisensor/scan/main.nf
Comment (grst) [inline software/msisensor/scan/main.nf:32]: Ha! found another stray echo ;)

**Review by grst — COMMENTED** [review]
Thanks for fixing the issues! 

Except for the two minor comments, this looks great now!

File: software/msisensor/msi/main.nf
Comment (kevbrick) [reply software/msisensor/msi/main.nf:22]: Yeah ... I discusssed this on slack a little with @drpatelh ... he suggested the single long tuple was best.

File: software/msisensor/scan/main.nf
Comment (kevbrick) [reply software/msisensor/scan/main.nf:32]: fixing...

File: software/msisensor/msi/main.nf
Comment (kevbrick) [reply software/msisensor/msi/main.nf:22]: I'll resolve for now, but can change if necessary ...

**Comment by kevbrick** [comment]
Strange. The modules pass all pytest tests locally but are failing all of them on gihub. Any idea why?

**Comment by kevbrick** [comment]
> Thanks for adding this!
> 
> Apart from the comments, can you please add an entry to the `pytest_software.yml` file - it is required to trigger the tests on the CI.

OK. Done ... tests are coming back good ...

File: software/msisensor/msi/main.nf
Comment (drpatelh) [inline software/msisensor/msi/main.nf:37]:
```suggestion
    output:
    tuple val(meta), path("${prefix}.output")         , emit: output
    tuple val(meta), path("${prefix}.output_dis")     , emit: output_dis
    tuple val(meta), path("${prefix}.output_germline"), emit: output_germline
    tuple val(meta), path("${prefix}.output_somatic") , emit: output_somatic
    path "*.version.txt"                              , emit: version

    script:
    def software = getSoftwareName(task.process)
    prefix       = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
    """
    msisensor \\
        msi \\
        -d $homopolymers \\
        -n $normal_bam \\
        -t $tumor_bam \\
        -o $prefix \\
        $options.args
```

File: software/msisensor/msi/main.nf
Comment (drpatelh) [inline software/msisensor/msi/main.nf:39]:
```suggestion
    echo \$(msisensor 2>&1) | sed -nE 's/Version:\\sv([0-9]\\.[0-9])/\\1/ p' > ${software}.version.txt
```

Comment (drpatelh) [reply software/msisensor/msi/main.nf:39]: Best to cancel out weird exit codes by using `echo` consistently.

File: software/msisensor/msi/main.nf
Comment (drpatelh) [reply software/msisensor/msi/main.nf:37]: Some re-structuring to output individual `emit` for each file created and to only use `$prefix` to name the output files. This just makes it more flexible for developer to call their output files whatever they want :) 

If you want to call the output files this for the tests then you can use something like the below in the `main.nf` for the tests (see `suffix` declaration at the end):
https://github.com/nf-core/modules/blob/13bd67ed8aca271a260c78ca41afe9e1f5c217d7/tests/software/gatk4/variantfiltration/main.nf#L5

File: software/msisensor/msi/meta.yml
Comment (drpatelh) [inline software/msisensor/msi/meta.yml:52]: All of the individual outputs needs to be listed here.

File: software/msisensor/scan/main.nf
Comment (drpatelh) [inline software/msisensor/scan/main.nf:33]:
```suggestion
                    -o ${prefix}.txt \\
```

File: software/msisensor/scan/main.nf
Comment (drpatelh) [inline software/msisensor/scan/main.nf:26]:
```suggestion
    tuple val(meta), path("*.txt"), emit: txt
    path "*.version.txt"          , emit: version
```

File: software/msisensor/scan/main.nf
Comment (drpatelh) [inline software/msisensor/scan/main.nf:36]:
```suggestion
    msisensor \\
        scan \\
        -d $fasta \\
        -o ${prefix}.txt \\
        $options.args

    echo \$(msisensor 2>&1) | sed -nE 's/Version:\\sv([0-9]\\.[0-9])/\\1/ p' > ${software}.version.txt
```

File: software/msisensor/scan/meta.yml
Comment (drpatelh) [inline software/msisensor/scan/meta.yml:36]: This should be `txt`?

**Review by drpatelh — COMMENTED** [review]
Thanks @kevbrick ! Looks great overall. Some minor adjustments required :)

**Comment by kevbrick** [comment]
> Thanks @kevbrick ! Looks great overall. Some minor adjustments required :)

Thanks for code review @drpatelh ... fixed all as suggested. There's still a requested change from @grst (1st code review) that seems to be blocking merge ... change was made but I can't seem to mark as resolved. Any suggestions?

**Review by grst — APPROVED** [review]
LGTM now

File: software/msisensor/msi/meta.yml
Comment (drpatelh) [inline software/msisensor/msi/meta.yml:52]:
```suggestion
  - output:
```

Comment (drpatelh) [reply software/msisensor/msi/meta.yml:52]: These need to the name `emit` channel e.g. `output`, `output_dis` etc

File: software/msisensor/msi/meta.yml
Comment (drpatelh) [inline software/msisensor/msi/meta.yml:56]:
```suggestion
  - output_dis:
```

File: software/msisensor/scan/main.nf
Comment (drpatelh) [inline software/msisensor/scan/main.nf:35]:
```suggestion
        -o ${prefix}.tab \\
```

File: software/msisensor/scan/main.nf
Comment (drpatelh) [inline software/msisensor/scan/main.nf:26]:
```suggestion
    tuple val(meta), path("*.tab"), emit: txt
    path "*.version.txt"          , emit: version
```

File: software/msisensor/scan/meta.yml
Comment (drpatelh) [inline software/msisensor/scan/meta.yml:36]:
```suggestion
  - tab:
```

File: software/msisensor/scan/meta.yml
Comment (drpatelh) [inline software/msisensor/scan/meta.yml:39]:
```suggestion
      pattern: "*.tab"
```

File: tests/software/msisensor/msi/main.nf
Comment (drpatelh) [inline tests/software/msisensor/msi/main.nf:13]:
```suggestion
    scaninput = [ [ id:'test', single_end:false ], // meta map
                  file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true) 
                ]
    MSISENSOR_SCAN ( scaninput )
```

File: tests/software/msisensor/msi/main.nf
Comment (drpatelh) [inline tests/software/msisensor/msi/main.nf:30]:
```suggestion
    Channel
        .from(
            [ [ id:'test', single_end:false ], // meta map
              file(params.test_data['sarscov2']['illumina']['test_methylated_paired_end_sorted_bam'], checkIfExists: true),
              file(params.test_data['sarscov2']['illumina']['test_methylated_paired_end_sorted_bam_bai'], checkIfExists: true),
              file(params.test_data['sarscov2']['illumina']['test_paired_end_sorted_bam'], checkIfExists: true),
              file(params.test_data['sarscov2']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExists: true) 
            ])
        .join(MSISENSOR_SCAN.txt)
        .collect()
        .set { input }
    
    MSISENSOR_MSI ( input )
```

Comment (drpatelh) [reply tests/software/msisensor/msi/main.nf:30]: What about something like that instead?

File: tests/software/msisensor/scan/main.nf
Comment (drpatelh) [inline tests/software/msisensor/scan/main.nf:11]:
```suggestion
    input = [ [ id:'test', single_end:false ], // meta map
              file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
            ]
```

**Comment by drpatelh** [comment]
Spotted a few things that still need to be changed @kevbrick ! Almost there!

**Comment by kevbrick** [comment]
> Spotted a few things that still need to be changed @kevbrick ! Almost there!
Oops ... just merged ... should I revert?

**Comment by drpatelh** [comment]
Ah, no worries 🙂  Maybe easier to just update in a follow up PR?

**Comment by kevbrick** [comment]
> Ah, no worries  Maybe easier to just update in a follow up PR?

:+1: Thanks for all the help ...

**Pull request merged**
